### PR TITLE
fix: deduplicate dependencies by artifact target

### PR DESCRIPTION
**Pull request by rvolosatovs**
### What does this PR try to resolve?

In cases when a compile target is specified for a bindep and the crate depending on it, cargo fails to deduplicate the crate dependencies and attempts to build the dependent crate only once with non-deterministic feature set, which breaks e.g. https://github.com/rvolosatovs/musl-bindep-feature-bug

Fix the issue by including the optional artifact compile target in the `Unit` in order to avoid wrongfully deduplicating the dependent crates

Fixes https://github.com/rust-lang/cargo/issues/11463 
Fixes https://github.com/rust-lang/cargo/issues/10837
Fixes https://github.com/rust-lang/cargo/issues/10525

Note, that this issue is already accounted for by `cargo`, but in different context a similar situation can occur while building the build script, which:
1. may be built for different target than the actual package target
2. may contain dependencies with different feature sets than the same dependencies in the dependency graph of the package itself

That's why this PR is simply reusing the existing functionality for deduplication

### How should we test and review this PR?

Build https://github.com/rvolosatovs/musl-bindep-feature-bug

### Additional information

This is based on analysis by @weihanglo in https://github.com/rust-lang/cargo/issues/10837#issuecomment-1339365374
I experimented with adding the whole `UnitFor` to the internal unit struct, but that seems unnecessary.

It would probably be nicer to refactor `IsArtifact` and instead turn it into a 3-variant enum with a possible compile target, but I decided against that to minimize the diff. Perhaps it's worth a follow-up?

### CODE REVIEW

**Comment by rvolosatovs** [comment]
cc @bstrie @weihanglo

**Comment by npmccallum** [comment]
FYI: @Byron

**Comment by weihanglo** [comment]
Byron probably is working on gitoxide integration. I will take a look then :)

r? @weihanglo 

BTW, could we have some tests for this patch?

**Comment by rvolosatovs** [comment]
> BTW, could we have some tests for this patch?

@weihanglo I'd love to do that, but I was having troubles coming up with a minimal reproduction of this issue, it works (as in, fails) with `rsa` crate, but that's probably not something we want to depend upon in tests. Do you have some clever ideas on how to trigger this bug by using just local crates.
I'll try to do a bit more exploration

Do you agree with the approach overall?

**Comment by rvolosatovs** [comment]
Update: I took my time to untangle the `rsa->num_bigint_dig->rand->rand_core` dependency chain and managed to reproduce the bug by only using local crates (based on this dependency chain) https://github.com/rvolosatovs/musl-bindep-feature-bug/commit/eab0536f8c7111617532838d38350a247cfbdf2d
I will proceed with adding a test case based on this. For now it's hard for me to wrap my head around it to simplify (if possible at all)

**Comment by rvolosatovs** [comment]
I managed to reproduce the error in https://github.com/rvolosatovs/cargo/commit/440517beeaac9a5e115ddee0f30f47aaf64ba544, which the fix is conveniently based on top of, the test fails on that commit and is fixed by the child commit.
I tried a few things, but unable to simplify the testcase further. Note that `edition =` is important and removing it at the top-level eliminates the issue

**Comment by weihanglo** [comment]
Sorry I haven't yet got time to do a full review on this. Just note that I think the bug isn't related to workspace, so maybe use path dependencies to make the gist of the test case clearer?
Besides, for #10837 `{ …, lib = ture, target = "<target>" }` it is worth one test as well. Thank you!

**Comment by rvolosatovs** [comment]
> Sorry I haven't yet got time to do a full review on this. Just note that I think the bug isn't related to workspace, so maybe use path dependencies to make the gist of the test case clearer? Besides, for #10837 `{ …, lib = ture, target = "<target>" }` it is worth one test as well. Thank you!

@weihanglo I've rebased on top of latest `master`, added a test case with `lib = true` based on example from #10837 and updated the doc in a few places.

You should be able to confirm both test case failures on `HEAD^` of this PR

**Comment by bstrie** [comment]
I have a branch at https://github.com/bstrie/cargo/tree/fix/bindeps-target that adds regression tests for both #10525 and #10837. I can confirm that these tests both fail on the current head, and both succeed on top of this PR. @rvolosatovs , can you pull this branch and add it to this PR?

**Comment by bstrie** [comment]
On further inspection it looks like the test that was added yesterday covers the #10837 case. I've updated my branch to only add a test for #10525.

File: src/cargo/core/compiler/unit_dependencies.rs
Comment (bstrie) [inline src/cargo/core/compiler/unit_dependencies.rs:888]: When used as part of a `let`, I'd suggest `match` as opposed to an `if-let-else`.

**Review by bstrie — COMMENTED** [review]
This looks good to me, I only have minor style and documentation comments. I'm not an authority on Cargo's codebase, but this looks like a low-risk addition that is unlikely to affect anyone that isn't using bindeps. @weihanglo , do you agree that this addresses what you described in your comment at https://github.com/rust-lang/cargo/issues/10837#issuecomment-1339365374 ? This patch is mercifully only 60 logical lines excluding the tests, so I hope that it is not too much burden on you to review. I have confirmed that the test cases as they exist satisfactorily address #10837, and once my own branch has been merged it will address #10525 as well.

**Comment by bstrie** [comment]
@weihanglo I had a style question about test cases, in my branch where I add the regression test for #10525 I copied the code verbatim from the issue, using the mock crates I listed there, but would you prefer that I "normalize" them to all the same dummy names used in the other test cases?

File: src/cargo/core/compiler/unit.rs
Comment (rvolosatovs) [reply src/cargo/core/compiler/unit.rs:79]: Yes, this comment is duplicated verbatim from https://github.com/rust-lang/cargo/blob/4534dd67bf7ceca6088a170f3f68a5bc9a37d437/src/cargo/core/profiles.rs#L843-L849

See, for example, https://github.com/rust-lang/cargo/blob/4534dd67bf7ceca6088a170f3f68a5bc9a37d437/src/cargo/core/profiles.rs#L1011-L1021 or https://github.com/rust-lang/cargo/blob/4534dd67bf7ceca6088a170f3f68a5bc9a37d437/src/cargo/core/profiles.rs#L1005-L1009 for usage

The actual value for this field in `Unit` originates from `UnitFor`

**Comment by rvolosatovs** [comment]
I've integrated @bstrie's test case for #10525 in https://github.com/rust-lang/cargo/pull/11478/commits/d9959b720d873abb86391cd5f26c2d40bab4ebd6 @weihanglo PTAL

**Review by bstrie — APPROVED** [review]

**Comment by bstrie** [comment]
LGTM

File: tests/testsuite/artifact_dep.rs
Comment (weihanglo) [inline tests/testsuite/artifact_dep.rs:2455]:
```suggestion
    // See https://github.com/rust-lang/cargo/issues/11463
```

File: tests/testsuite/artifact_dep.rs
Comment (weihanglo) [inline tests/testsuite/artifact_dep.rs:2677]: Maybe a descriptive title.

And other tests should be `decouple_from` instead of `deduplication`, no?

```suggestion
fn same_target_transitive_dependency_decouple_from_hostdep() {
    // See https://github.com/rust-lang/cargo/issues/10525
```

File: src/cargo/ops/cargo_compile/mod.rs
Comment (weihanglo) [inline src/cargo/ops/cargo_compile/mod.rs:438]: (nit)
 
```suggestion 
            .any(|(unit, _)| unit.artifact_target_for_features.is_some())
```

File: src/cargo/ops/cargo_compile/mod.rs
Comment (weihanglo) [inline src/cargo/ops/cargo_compile/mod.rs:577]: If I understand change here correctly, it strives to share artifact deps with other same deps for the same target, right? I'd recommend this having its own commit. And maybe need a test to verify this behaviour?

**Review by weihanglo — COMMENTED** [review]
Looks reasonable to me. Thank you for posting this fix!

File: src/cargo/ops/cargo_compile/mod.rs
Comment (weihanglo) [inline src/cargo/ops/cargo_compile/mod.rs:650]: Could we have a simple comment explaining why a `None` being passed?

Comment (rvolosatovs) [reply src/cargo/ops/cargo_compile/mod.rs:649]: Done

File: tests/testsuite/artifact_dep.rs
Comment (rvolosatovs) [reply tests/testsuite/artifact_dep.rs]: Done. I simplified this test case quite a bit and it turned out that it has nothing to do with proc macros and seems to be a duplicate of #11463 #10837
I left the test in regardless, since it has a slightly different structure.

Note, that I uncovered a pretty bizarre inconsistency - if `e` crate is renamed to `a`, the issue does not occur in upstream. How is naming related here is beyond me, but it does seem that there are more things broken the resolver

cc @bstrie

File: src/cargo/ops/cargo_compile/mod.rs
Comment (rvolosatovs) [reply src/cargo/ops/cargo_compile/mod.rs]: During the construction of `unit_graph`, `artifact_target_for_features` was set to decouple the dependencies, in this step, duplicates are merged. I am not sure a separate commit makes sense here, since this is only relevant for artifact dependencies with `artifact_target_for_features` set, so I could have a single commit, which breaks artifact deps feature, since cargo would fail to compile such packages with errors about duplicate paths and a follow-up fixing the issue, but I don't see why would that be preferrable

**Comment by rvolosatovs** [comment]
@weihanglo I believe all comments are addressed.
Note, that I added the new field to `Debug` implementation, which I missed earlier
I also refactored #10525 test case significantly and found out that it's unrelated to proc macros, but uncovered a concerning naming-related inconsistency https://github.com/rust-lang/cargo/pull/11478#discussion_r1052550831

File: tests/testsuite/artifact_dep.rs
Comment (bstrie) [reply tests/testsuite/artifact_dep.rs]: Part of my rationale in leaving in the proc macros was to exercise different codepaths as compared to the other test. In addition, while this test case is greatly simplified compared to the original reproducer, care was taken to not simplify it *too* much, in order to produce the same error message as the original maximized bug. I don't mind if it gets simplified further, although this runs the risk of reducing the value of this test as a regression test. I leave the judgment here up to @weihanglo as to whether or not this test is valuable to have, and whether or not it should exercise the proc macro machinery.

**Review by weihanglo — COMMENTED** [review]
Thank you for your update. This looks really nice. We're getting closer to merge!

---

> but uncovered a concerning naming-related inconsistency [#11478 (comment)](https://github.com/rust-lang/cargo/pull/11478#discussion_r1052550831) , which could be related to lexicographic sorting of the dependencies in some way.

I believe it's actually the same issue. [This line] makes Cargo skip redundant unit computation. Before this pull request, either with package named `a` or `e` cannot help discriminate the difference of package `c` being built for artifact dep or normal dep.

<details>

<summary>Expand to read the detail</summary>

Below is how `unit_dependencies` are computed with the package named `a`.

```ocaml
("foo", [])
    -> unit_deps: [("a", []), ("bar", [])]
("a", [])
    -> unit_deps: [("c", []), ("d", ["feature"])]
("c", [])
    -> unit_deps: [("d", ["feature"])]
("d", ["feature"])
    -> unit_deps: []
("d", ["feature"])  # cache hit, skipping...
("bar", [])
    -> unit_deps: [("b", [])]
("b", [])
    -> unit_deps: [("c", [])]
("c", []) # cache hit, skipping...
```

You can see that `UnitDep`s of `a` are computed first, so we get only one `c`, which depends on `d` with `feature` enabled. Therefore, when we compute unit deps of `b`, cache hit for `c`. so they reuse `d` with `feature` enabled for the whole dependency graph.

If we change package name `a` to `e`:

```ocaml
("foo", [])
    -> unit_deps: [("bar", []), ("e", [])]
("bar", [])
    -> unit_deps: [("b", [])]
("b", [])
    -> unit_deps: [("c", [])]
("c", [])
    -> unit_deps: [("d", [])]
("d", [])
    -> unit_deps: []
("e", [])
    -> unit_deps: [("c", []), ("d", ["feature"])]
("c", []) # cache hit, skipping...
("d", ["feature"])
    -> unit_deps: []
```

Cargo now computes `bar` first, and get a `c` with d without any feature. Next when computing deps of `e`, Cargo skips and reuses preivous `c`. That leads to havning two different versions of `d` in the subgraph of `e`. (One with `feature` enabled and one without). And  that confuses rustc.


</details>

[This line]: https://github.com/rust-lang/cargo/blob/74c7aab19a9b3a045e7af13319409a9de2cf4ef7/src/cargo/core/compiler/unit_dependencies.rs#L247

File: src/cargo/ops/cargo_compile/mod.rs
Comment (rvolosatovs) [reply src/cargo/ops/cargo_compile/mod.rs]: It indeed looks like this case is not covered by existing test suite, but the resulting compiler fails on real-word codebase with a wall of errors like these:
```
warning: output filename collision.
The lib target `webpki` in package `webpki v0.22.0` has the same output filename as the lib target `webpki` in package `webpki v0.22.0`.
Colliding filename is: /home/rvolosatovs/src/github.com/enarx/enarx/target/x86_64-unknown-linux-musl/debug/deps/libwebpki-19484ee9a1315a74.rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The lib target `webpki-roots` in package `webpki-roots v0.22.4` has the same output filename as the lib target `webpki-roots` in package `webpki-roots v0.22.4`.
Colliding filename is: /home/rvolosatovs/src/github.com/enarx/enarx/target/x86_64-unknown-linux-musl/debug/deps/libwebpki_roots-f41f1a48aec910b8.rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The lib target `webpki-roots` in package `webpki-roots v0.22.4` has the same output filename as the lib target `webpki-roots` in package `webpki-roots v0.22.4`.
Colliding filename is: /home/rvolosatovs/src/github.com/enarx/enarx/target/x86_64-unknown-linux-musl/debug/deps/libwebpki_roots-f41f1a48aec910b8.rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The lib target `witx` in package `witx v0.9.1 (https://github.com/bytecodealliance/wasmtime?rev=6f50ddaaf2ab8205b6e850361dd2cc662819f431#6f50ddaa)` has the same output filename as the lib target `witx` in package `witx v0.9.1 (https://github.com/bytecodealliance/wasmtime?rev=6f50ddaaf2ab8205b6e850361dd2cc662819f431#6f50ddaa)`.
Colliding filename is: /home/rvolosatovs/src/github.com/enarx/enarx/target/debug/deps/libwitx-158c96bbbf8bb0ec.rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The lib target `witx` in package `witx v0.9.1 (https://github.com/bytecodealliance/wasmtime?rev=6f50ddaaf2ab8205b6e850361dd2cc662819f431#6f50ddaa)` has the same output filename as the lib target `witx` in package `witx v0.9.1 (https://github.com/bytecodealliance/wasmtime?rev=6f50ddaaf2ab8205b6e850361dd2cc662819f431#6f50ddaa)`.
Colliding filename is: /home/rvolosatovs/src/github.com/enarx/enarx/target/debug/deps/libwitx-158c96bbbf8bb0ec.rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
thread 'main' panicked at 'build script output collision for proc-macro2 v1.0.43/c9c8981de1417d5d
old=BuildOutput { library_paths: [], library_links: [], linker_args: [], cfgs: ["use_proc_macro", "wrap_proc_macro", "proc_macro_span"], check_cfgs: [], env: [], metadata: [], rerun_if_changed: ["build.rs"], rerun_if_env_changed: [], warnings: [] }
new=BuildOutput { library_paths: [], library_links: [], linker_args: [], cfgs: ["use_proc_macro", "wrap_proc_macro", "proc_macro_span"], check_cfgs: [], env: [], metadata: [], rerun_if_changed: ["build.rs"], rerun_if_env_changed: [], warnings: [] }', src/cargo/core/compiler/custom_build.rs:1010:39
```

I'll work on adding a test case

Comment (rvolosatovs) [reply src/cargo/ops/cargo_compile/mod.rs]: See https://github.com/rust-lang/cargo/pull/11478/commits/548b2528fb1a1c725a5aef14c80d9a401deb3ae9

I've now split this into two commits, first commit basically introducing the bug, next one adding test reproducing it and then the fix

**Review by weihanglo — APPROVED** [review]
Looks really great! Both of the commit history and patch are quite clear. Thank you @rvolosatovs for working on this patch. Also thanks @bstrie for taking care of artifact dependencies for a long while. I might not be a good communicator but still happy to collaborate with you :)

The only concern I have is `artifact_target_for_features` being duplicated. It is not that bad as we already added a link to `UnitFor::artifact_target_for_features`. If @Byron or someone from the Cargo team feels it is wrong, please call it out. Thanks!

**Pull request merged**
